### PR TITLE
[PLAY-88] Currency kit new abbreviate prop

### PR DESCRIPTION
**Pull request by kyleschechter**
#### Screens

[INSERT SCREENSHOT]

#### Breaking Changes

[Yes/No (Explain)]

#### Runway Ticket URL

[https://nitro.powerhrg.com/runway/backlog_items/PLAY-88](url)

#### How to test this

[INSERT TESTING DETAILS]

#### Checklist:

- [ ] **LABELS** Add a label: `enhancement`, `bug`, `improvement`, `new kit`, `deprecated`, or `breaking`. See [Changelog & Labels](https://github.com/powerhome/playbook/wiki/Changelog-&-Labels) for details.
- [ ] **DEPLOY** Please add the `Milano` label when you are ready for a review.
- [ ] **SCREENSHOT** Please add a screen shot or two.
- [ ] **SPECS** Please cover your changes with specs.
- [ ] **READ DOCS** Please make sure you have read and understand the [Playbook Release Process](https://github.com/powerhome/playbook/wiki/Playbook-Releases)

### CODE REVIEW

File: playbook/app/pb_kits/playbook/pb_currency/currency.rb [hunk 1]
```diff
@@ -84,10 +87,17 @@ def whole_value
...
+        number_to_human(value, units: { thousand: "K", million: "M", billion: "B", trillion: "T" })
+      end
+
       def units_element
         _, decimal_part = amount.split(".")
         if unit.nil?
-          decimal_part.nil? ? ".00" : ".#{decimal_part}"
+          if abbreviate == false
+            decimal_part.nil? ? ".00" : ".#{decimal_part}"
+          end
```
Comment (JelenaD1) [inline playbook/app/pb_kits/playbook/pb_currency/currency.rb:100]: I'm wondering if we can optimize this and say `if unit.nil? && abbreviate == false` ?

File: playbook/spec/pb_kits/playbook/pb_currency/currency_spec.rb [hunk 2]
```diff
@@ -13,6 +13,7 @@
   it { is_expected.to define_enum_prop(:size).with_default("md").with_values("sm", "md", "lg") }
   it { is_expected.to define_prop(:symbol).with_default("$").of_type(Playbook::Props::String) }
   it { is_expected.to define_enum_prop(:variant).with_default("default").with_values("default", "light", "bold") }
+  it { is_expected.to define_prop(:abbreviate).of_type(Playbook::Props::Boolean) }
```
Comment (JelenaD1) [inline playbook/spec/pb_kits/playbook/pb_currency/currency_spec.rb:16]: Maybe will worth it to test the default as well in case someone change it in future.

File: playbook/app/pb_kits/playbook/pb_currency/currency.rb (hunk 1 above)
Comment (kyleschechter) [reply playbook/app/pb_kits/playbook/pb_currency/currency.rb:100]: I'll try it out

File: playbook/spec/pb_kits/playbook/pb_currency/currency_spec.rb (hunk 2 above)
Comment (kyleschechter) [reply playbook/spec/pb_kits/playbook/pb_currency/currency_spec.rb:16]: I'll try that too

File: playbook/app/pb_kits/playbook/pb_currency/currency.rb (hunk 1 above)
Comment (kyleschechter) [reply playbook/app/pb_kits/playbook/pb_currency/currency.rb]: done

File: playbook/spec/pb_kits/playbook/pb_currency/currency_spec.rb (hunk 2 above)
Comment (kyleschechter) [reply playbook/spec/pb_kits/playbook/pb_currency/currency_spec.rb]: done

**Review by JelenaD1 — APPROVED** [review]

File: playbook/app/pb_kits/playbook/pb_currency/_currency.jsx [hunk 3]
```diff
@@ -69,6 +71,20 @@ const Currency = (props: CurrencyProps) => {
...
+      maximumFractionDigits: 1,
+    }).format(whole.split(',').join('')).toString()
+    let output
+    if (arg === 'amount') {
+      output = num.slice(0, -1)
+    } else if (arg === 'unit') {
+      output = num.slice(-1)
+    }
+    return output
+  }
```
Comment (thestephenmarshall) [inline playbook/app/pb_kits/playbook/pb_currency/_currency.jsx:86]:
```suggestion
  const getFormattedNumber = (input) => new Intl.NumberFormat('en-US', {
      notation: 'compact',
      maximumFractionDigits: 1,
    }).format(input)
  
  type AbbrType = 'amount' | 'unit'
  
  const getAbbreviatedValue = (abbrType: AbbrType) => {
    const num = `${getFormattedNumber(whole.split(',').join(''))}`,
          isAmount = abbrType === 'amount',
          isUnit =  abbrType === 'unit'
               
    return isAmount ? num.slice(0, -1) : isUnit ? num.slice(-1) : ''
  }
```

A couple of things about the original code here:

- Lower readability 
- The number formatter should be abstracted, especially to allow future re-use. Doing so also helps readability
  - Functions should strive to do one thing well and preferably only one
- `arg` is way to generic for *any* use case. Name your arguments contextually and don't forget to specify types. Now we can tell exactly the values that can be accepted. 
- The function was not returning any value if neither enumerated value was provided. It would just be `undefined` which would be bad

File: playbook/app/pb_kits/playbook/pb_currency/_currency.jsx [hunk 4]
```diff
@@ -92,18 +108,19 @@ const Currency = (props: CurrencyProps) => {
             dark={dark}
             size={sizes[size]}
         >
-          {`${whole}`}
+          {abbreviate ? getAbbreviatedValue('amount') : whole}
```
Comment (thestephenmarshall) [inline playbook/app/pb_kits/playbook/pb_currency/_currency.jsx:111]: We should steer towards leaving logic out of the templating area. Could you please move this to a local variable?

File: playbook/app/pb_kits/playbook/pb_currency/_currency.jsx [hunk 5]
```diff
@@ -92,18 +108,19 @@ const Currency = (props: CurrencyProps) => {
...
             className="unit"
             color="light"
             dark={dark}
         >
+          {abbreviate ? getAbbreviatedValue('unit') : null}
           <If condition={unit}>
             {unit}
             <Else />
-            {`.${decimal}`}
+            {abbreviate ? '' : `.${decimal}`}
```
Comment (thestephenmarshall) [inline playbook/app/pb_kits/playbook/pb_currency/_currency.jsx:123]: Same as the above comment ^

**Review by thestephenmarshall — COMMENTED** [review]
Coming along. Got some nits to be addressed.

File: playbook/app/pb_kits/playbook/pb_currency/_currency.jsx [hunk 6]
```diff
@@ -69,6 +71,24 @@ const Currency = (props: CurrencyProps) => {
...
+  const getAbbreviatedValue = (abbrType: AbbrType) => {
+    const num = `${getFormattedNumber(whole.split(',').join(''))}`,
+      isAmount = abbrType === 'amount',
+      isUnit =  abbrType === 'unit'
+    return isAmount ? num.slice(0, -1) : isUnit ? num.slice(-1) : ''
+  }
+
+  const getAmount = abbreviate ? getAbbreviatedValue('amount') : whole,
+    getAbbreviation = abbreviate ? getAbbreviatedValue('unit') : null,
+    getDecimalValue = abbreviate ? '' : `.${decimal}`
```
Comment (JelenaD1) [inline playbook/app/pb_kits/playbook/pb_currency/_currency.jsx:90]: I'm curious why you don't  define a variable with a `const` keyword for last two examples?

File: playbook/spec/pb_kits/playbook/pb_currency/currency_spec.rb [hunk 7]
```diff
@@ -13,6 +13,7 @@
   it { is_expected.to define_enum_prop(:size).with_default("md").with_values("sm", "md", "lg") }
   it { is_expected.to define_prop(:symbol).with_default("$").of_type(Playbook::Props::String) }
   it { is_expected.to define_enum_prop(:variant).with_default("default").with_values("default", "light", "bold") }
+  it { is_expected.to define_prop(:abbreviate).with_default(false).of_type(Playbook::Props::Boolean) }
```
Comment (JelenaD1) [inline playbook/spec/pb_kits/playbook/pb_currency/currency_spec.rb:16]: Do we expect any change in a classname when `abbreviate` prop is being passed?

File: playbook/app/pb_kits/playbook/pb_currency/_currency.jsx (hunk 6 above)
Comment (kyleschechter) [reply playbook/app/pb_kits/playbook/pb_currency/_currency.jsx:90]: I actually saw this done a couple times in other files. When you have a bunch of `const`s one after the other, you can separate them by commas and old put one `const` before the first one @JelenaD1

Comment (kyleschechter) [reply playbook/app/pb_kits/playbook/pb_currency/_currency.jsx:90]: *only

File: playbook/spec/pb_kits/playbook/pb_currency/currency_spec.rb (hunk 7 above)
Comment (kyleschechter) [reply playbook/spec/pb_kits/playbook/pb_currency/currency_spec.rb:16]: good question. No, because the only thing that is changing is the values of the variables being displayed in the return. The class name will not change

File: playbook/app/pb_kits/playbook/pb_currency/_currency.jsx (hunk 6 above)
Comment (JelenaD1) [reply playbook/app/pb_kits/playbook/pb_currency/_currency.jsx:90]: I did not know that. Thanks for explaining.

File: playbook/spec/pb_kits/playbook/pb_currency/currency_spec.rb (hunk 7 above)
Comment (kyleschechter) [reply playbook/spec/pb_kits/playbook/pb_currency/currency_spec.rb:16]: Let me rephrase that 'no', I don't think we can expect any change

File: playbook/app/pb_kits/playbook/pb_currency/_currency.jsx (hunk 6 above)
Comment (thestephenmarshall) [reply playbook/app/pb_kits/playbook/pb_currency/_currency.jsx:90]: Mostly just to keep word trash down in the code. Seeing `const` a bunch of times (unnecessarily) is going to reduce readability and make the code appear more monotonous. What is best about this approach is you can now visually group variables of concern together.

File: playbook/spec/pb_kits/playbook/pb_currency/currency_spec.rb (hunk 7 above)
Comment (thestephenmarshall) [reply playbook/spec/pb_kits/playbook/pb_currency/currency_spec.rb:16]: I am also looking at the Jest test in place. I think there are some necessary additions.

**Review by thestephenmarshall — CHANGES_REQUESTED** [review]
@kyleschechter I see that `currency_spec.rb` and `currency.test.js` are testing two different scenarios. We want to ensure that both the Ruby and the React versions of each test are testing the same cases respectively.
